### ISSUE TITLE
Hinter doesn't clear line correctly

### ISSUE THREAD

**Issue by kpcyrd** [kkawakam/rustyline#149]
It seems the hinter isn't cleared when it doesn't match anymore (hint returns None):
```
% cargo run --example example
    Finished dev [unoptimized + debuginfo] target(s) in 0.05s
     Running `target/debug/examples/example`
>> helloxyorld
```

Steps to reproduce are typing `hello`, this would hint ` World`, then type `xy` so the hint doesn't match anymore, but the old hint is still displayed until the line is cleared.

**Comment by gwenn** [comment]
`linenoise`/`replxx` do not have this bug because the optim is disabled when an hint callback is provided:
https://github.com/antirez/linenoise/blob/master/linenoise.c#L648
https://github.com/AmokHuginnsson/replxx/blob/master/src/inputbuffer.cxx#L1176

**Comment by gwenn** [comment]
```patch
Index: src/edit.rs
===================================================================
--- src/edit.rs	(revision 17ae324d6d9ff7cc1f200eec73d8d9e831f8bb26)
+++ src/edit.rs	(date 1536411581000)
@@ -32,6 +32,7 @@
     pub changes: Rc<RefCell<Changeset>>, // changes to line, for undo/redo
     pub hinter: Option<&'out Hinter>,
     pub highlighter: Option<&'out Highlighter>,
+    no_hint: bool, // `false` if an hint has been displayed
 }
 
 impl<'out, 'prompt> State<'out, 'prompt> {
@@ -57,6 +58,7 @@
             changes: Rc::new(RefCell::new(Changeset::new())),
             hinter,
             highlighter,
+            no_hint: true,
         }
     }
 
@@ -130,10 +132,12 @@
         Ok(())
     }
 
-    fn hint(&self) -> Option<String> {
+    fn hint(&mut self) -> Option<String> {
         if let Some(hinter) = self.hinter {
+            self.no_hint = false;
             hinter.hint(self.line.as_str(), self.line.pos())
         } else {
+            self.no_hint = true;
             None
         }
     }
@@ -186,10 +190,11 @@
         if let Some(push) = self.line.insert(ch, n) {
             if push {
                 let prompt_size = self.prompt_size;
+                let no_hint = self.no_hint;
                 let hint = self.hint();
                 if n == 1
                     && self.cursor.col + ch.width().unwrap_or(0) < self.out.get_columns()
-                    && hint.is_none() // TODO refresh only current line
+                    && (hint.is_none() && no_hint) // TODO refresh only current line
                     && !self.highlighter.map_or(true, |h| h.highlight_char(ch.encode_utf8(&mut self.byte_buffer)))
                 {
                     // Avoid a full update of the line in the trivial case.
```

**Comment by gwenn** [comment]
Should be fixed with the 2.1.0 release

**Comment by kpcyrd** [comment]
Thanks for the fast fix! :)
